### PR TITLE
Use ud asset mirror inside metadata service

### PR DESCRIPTION
**Pull request by markvital**
PFP NFT image should be cached on the side metadata service. 
After the event when 'social.picture.value' is changed on the blockchain and while putting in metadata service mirror, the picture should be retrieved and put in GCS CDN.
When the user retrieves image using '/image/:domainOrToken' or '/images-srt/:domainOrToken' we first check if the image is present in CDN and try to retrieve it.

TODO:

- [x] - support both _withOverlay=true_ and false (now only withOverlay=false is working)
- [ ] - add bigger resolution (1024x1024)
- [x] - refactor NFT metadata retrieval code, if needed
- [x] - remove comments and clean the code

### Related issue
https://linear.app/unstoppable-domains/issue/ECUX-276/use-ud-asset-mirror-inside-metadata-service

### CODE REVIEW

**Review by iurevych — CHANGES_REQUESTED** [review]
Feel free to re-request review whenever we can continue with QA. Currently uploads into GCP aren't working.

File: src/workers/eth/EthUpdater.ts [hunk 1]
```diff
@@ -258,6 +259,10 @@ export class EthUpdater {
     resolution.resolution[key] = value;
     domain.setResolution(resolution);
     await domainRepository.save(domain);
+    if (key === 'social.picture.value' && !!value) {
+      // temp. await for debugging. TODO: remove await, make it asynchronous
+      await cacheSocialPictureInCDN(value, domain, resolution);
```
Comment (DancingAxolotl) [inline src/workers/eth/EthUpdater.ts:264]: Should we also do it for the `ZilUpdater`?

Comment (markvital) [reply src/workers/eth/EthUpdater.ts:264]: Sure, I'll add it!

File: src/env.ts [hunk 2]
```diff
@@ -240,4 +240,11 @@ export const env = {
   OPENSEA: {
     API_KEY: process.env.OPENSEA_API_KEY,
   },
+  CLOUD_STORAGE: {
+    CLIENT_ASSETS: {
+      BUCKET_ID:
+        process.env.CLOUD_STORAGE_BUCKET_ID || 'resolution-client-assets',
```
Comment (DancingAxolotl) [inline src/env.ts:246]: I might be missing something but shouldn't there be an API key for uploading files to the cache?

**Review by DancingAxolotl — COMMENTED** [review]
Left a couple comments, but overall looks good. I will defer to ecomm team for final approval.

File: src/utils/socialPicture/index.ts [hunk 3]
```diff
@@ -122,3 +131,100 @@ export const createSocialPictureImage = (
...
+  const hostname =
+    env.CLOUD_STORAGE.API_ENDPOINT_URL || 'https://storage.googleapis.com';
+  const bucket = storage.bucket(bucketName);
+
+  const [fileExists] = await bucket.file(fileName).exists();
+  if (fileExists) {
+    const contents = await bucket.file(fileName).download();
+    return contents.toString(); // maybe replace this with stream buffer in the future
+  } else {
+    return null;
```
[the remark this answers is not in the document]
Comment (markvital) [reply src/utils/socialPicture/index.ts:217]: If the image is not in CDN, we just show default placeholder for the domain.

For the old domains we'll do a migration separately, uploading images for all existent PFP NFT in CDN

File: src/env.ts (hunk 2 above)
Comment (markvital) [reply src/env.ts:246]: Only bucket name. As I understood from talking to devOps team, the API key from GCS is stored somewhere else.

File: src/workers/eth/EthUpdater.ts (hunk 1 above)
Comment (markvital) [reply src/workers/eth/EthUpdater.ts:263]: Talked to Sergii, decided we won't be supporting Zil domains

**Comment by markvital** [comment]
The tests that are failing are flaky: different test fails when I re-run the job. Ignoring them for now.
Screen Shot 2022-08-04 at 7 38 17 PM
Screen Shot 2022-08-04 at 6 59 11 PM

File: src/utils/socialPicture/index.ts [hunk 4]
```diff
@@ -106,19 +115,117 @@ export const createSocialPictureImage = (
...
+    parsePictureRecord(socialPic);
+  const nftPfpFolder = 'nft-pfp';
+  return `${nftPfpFolder}/${chainId}_${nftStandard}:${contractAddress}_${tokenId}.svg`;
+};
+
+export const getNFTFilenameWithOverlayInCDN = (socialPic: string): string => {
+  const { chainId, nftStandard, contractAddress, tokenId } =
+    parsePictureRecord(socialPic);
+  const nftPfpFolder = 'nft-pfp';
+  return `${nftPfpFolder}/${chainId}_${nftStandard}:${contractAddress}_${tokenId}_overlay.svg`;
```
Comment (iurevych) [inline src/utils/socialPicture/index.ts:136]: We can combine `getNFTFilenameInCDN` and `getNFTFilenameWithOverlayInCDN` into one function

File: src/utils/socialPicture/index.ts [hunk 5]
```diff
@@ -106,19 +115,117 @@ export const createSocialPictureImage = (
...
+): Promise<void> => {
+  const fileName = getNFTFilenameInCDN(socialPic);
+  const fileNameWithOverlay = getNFTFilenameWithOverlayInCDN(socialPic);
+  const bucketName = env.CLOUD_STORAGE.CLIENT_ASSETS.BUCKET_ID;
+  const bucket = storage.bucket(bucketName);
+
+  const [fileExists] = await bucket.file(fileName).exists();
+  const [fileWithOverlayExists] = await bucket
+    .file(fileNameWithOverlay)
+    .exists();
```
Comment (iurevych) [inline src/utils/socialPicture/index.ts:152]: We can use `Promise.all` here to parallelise requests.

File: src/controllers/MetaDataController.ts [hunk 6]
```diff
@@ -204,11 +199,16 @@ export class MetaDataController {
...
-      withOverlay,
-    );
+    const socialPictureValue = resolution.resolution['social.picture.value'];
+    const socialPicture =
+      socialPictureValue &&
+      (await getNftPfpImageFromCDN(socialPictureValue, withOverlay));
+
+    // we consider that NFT picture is verified if the picture is present in our CDN cache.
+    // It means it was verified before caching.
+    const isSocialPictureVerified = socialPicture
```
Comment (iurevych) [inline src/controllers/MetaDataController.ts:209]:
```suggestion
    const isSocialPictureVerified = Boolean(socialPicture);
```

**Review by iurevych — APPROVED** [review]
👍  awesome

**Comment by iurevych** [comment]
Release plan:
1. Upload historical NFT PFPs into GCP bucket
2. Release this PR

**Comment by iurevych** [comment]
@DancingAxolotl can we by any chance run a similar script against production metadata service instance in order to populate cached images into our bucket?

``` ts
const resolutions = await DomainsResolution.find({where: ,...});
const domains = await ...;

for (const resolution of resolutions) {
  const domain = domains.find(d => resolutions.domainId === d.id);
  await cacheSocialPictureInCDN(resolution.resolution['social.picture.value'], domain, resolution);
}
```

**Pull request merged**